File: staylio_frontend/common/services/booking.service.ts
import { PaginationResponse } from "@common/interfaces";
import { ApiResponse } from "@common/interfaces/ApiResponse";
import { BookingRequest } from "@common/interfaces/request/BookingRequest";
import { BookingStatusRequest } from "@common/interfaces/request/BookingStatusRequest";
import { BookingQueryParams, QueryParams } from "@common/interfaces/request/QueryParams";
import { BookingHistoryResponse } from "@common/interfaces/response/BookingHistoryResponse";
import { BookingPreviewResponse } from "@common/interfaces/response/BookingPreviewResponse";
import { BookingResponse } from "@common/interfaces/response/BookingResponse";
import { axiosInstance } from "@common/utils/axiosInstance";

// API lấy danh sách đơn đặt phòng
export const getAllBookings = async (params: BookingQueryParams): Promise<ApiResponse<PaginationResponse<BookingResponse>>> => {
    try {
        const response = await axiosInstance.get("/bookings", { params });
        return response.data;
    } catch (error) {
        console.error("Lấy danh sách thất bại!", error);
        throw error;
    }
}

// API lấy chi tiết 1 đơn đặt phòng
export const getBookingById = async (id: number): Promise<ApiResponse<BookingResponse>> => {
    try {
        const response = await axiosInstance.get(`/bookings/${id}`);
        return response.data;
    } catch (error) {
        console.error("Lấy chi tiết 1 đơn đặt phòng thất bại!", error);
        throw error;
    }
}

// API cập nhật trạng thái đơn phòng
export const updateStatus = async (id: number, data: BookingStatusRequest): Promise<ApiResponse<string>> => {
    try {
        const response = await axiosInstance.patch(`/bookings/${id}/status`, data);
        return response.data;
    } catch (error) {
        console.error("Cập nhật trạng thái thất bại", error);
        throw error;
    }
}

// API Preview giá phòng trc khi đặt
export const previewBooking = async (request: BookingRequest): Promise<ApiResponse<BookingPreviewResponse>> => {
    try {
        const response = await axiosInstance.post("/bookings/preview", request);
        return response.data;
    } catch (error) {
        console.error("Preview phòng thất bại", error);
        throw error;
    }
}

// API đặt phòng
export const createBooking = async (request: BookingRequest): Promise<ApiResponse<BookingResponse>> => {
    try {
        const response = await axiosInstance.post("/bookings", request);
        return response.data;
    } catch (error) {
        console.error("Đặt phòng thất bại", error);
        throw error;
    }
}

// API lấy các ngày đã đặt của một phòng
export const getBookedDates = async (roomId: number): Promise<ApiResponse<{ checkInDate: string; checkOutDate: string }[]>> => {
    try {
        const response = await axiosInstance.get(`/bookings/room/${roomId}/booked-dates`);
        return response.data;
    } catch (error) {
        console.error("Lấy danh sách ngày đã đặt thất bại", error);
        throw error;
    }
}

// API lấy lịch sử đặt phòng
export const getHistoryBookings = async (params: QueryParams): Promise<ApiResponse<PaginationResponse<BookingHistoryResponse>>> => {
    try {
        const response = await axiosInstance.get("/bookings/history", { params });
        return response.data;
    } catch (error) {
        console.error("Lấy danh sách lịch sử đặt phòng thất bại", error);
        throw error;
    }
}

File: staylio_backend/src/main/java/com/example/staylio_backend/service/BookingService.java
package com.example.staylio_backend.service;

import com.example.staylio_backend.config.security.principle.UserPrincipal;
import com.example.staylio_backend.dto.request.BookingFilterRequest;
import com.example.staylio_backend.dto.request.BookingHistoryRequest;
import com.example.staylio_backend.dto.request.BookingRequest;
import com.example.staylio_backend.dto.request.BookingStatusRequest;
import com.example.staylio_backend.dto.response.BookingHistoryResponse;
import com.example.staylio_backend.dto.response.BookingPreviewResponse;
import com.example.staylio_backend.dto.response.BookingResponse;
import com.example.staylio_backend.dto.response.DateRangeResponse;
import com.example.staylio_backend.dto.response.page.PaginationResponse;

import java.util.List;

public interface BookingService {
    PaginationResponse<BookingResponse> getAllBookings(BookingFilterRequest request, UserPrincipal principal);

    BookingResponse getBookingById(Long id, UserPrincipal principal);

    void updateStatus(Long id, BookingStatusRequest request, UserPrincipal principal);

    BookingPreviewResponse previewBooking(BookingRequest request, UserPrincipal principal);

    BookingResponse createBooking(BookingRequest request, UserPrincipal principal);

    List<DateRangeResponse> getBookedDates(Long roomId);

    PaginationResponse<BookingHistoryResponse> getMyBookings(
            BookingHistoryRequest request,
            UserPrincipal principal
    );
}
